### PR TITLE
docs: Adding brief documentation to coreComponent packages

### LINKED ISSUE #3292: Improved documentation for `coreComponents` packages
**What needs to be documented?**
Many `coreComponents` packages such as `codingUtilities`, `common`, etc.. seems to not be documented on what they contain, manage, should not contain. As a developper, it is currently not obvious the responsibilities of some of these packages.

**What sort of documentation would help?**
DevGuide doc.

**What are the deficiencies of the existing documentation?**
The files and classes are documented but the containing packages lack description.

**Additional notes or context**
In addition, it might be very useful to have a small documentation at the top of each `CMakeLists.txt` file such that brief info about the package is accessible easily, especially while making modification in the package.

### PR DESCRIPTION
**Pull request by paloma-martinez**
This PR aims at providing a small documentation of `coreComponents` packages for developers, following the issue #3292.

A quick look at the CMakeList.txt will give indications on what the package contains.

### CODE REVIEW

File: src/coreComponents/codingUtilities/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/codingUtilities/CMakeLists.txt]: @paloma-martinez If you use decorate your headers with these line separators, can you use a consistent length of, let's say, 100 characters?

Comment (paloma-martinez) [reply src/coreComponents/codingUtilities/CMakeLists.txt]: Ok, I'll include it in the next modifications.

File: src/coreComponents/codingUtilities/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/codingUtilities/CMakeLists.txt]:
```suggestion
# Contains definition of runtime types. Also contains common traits.
```

File: src/coreComponents/common/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/common/CMakeLists.txt]: What do you think of:
```suggestion
# Contains the definition of the basic static types of GEOS.
# Also provides commonly used components for such as logging, formatting, memory and wrapper of dependencies.
```

File: src/coreComponents/constitutiveDrivers/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/constitutiveDrivers/CMakeLists.txt]: Open & quite naive question: Does the notion of `driver` would need to be defined here?

File: src/coreComponents/dataRepository/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/dataRepository/CMakeLists.txt]:
```suggestion
# Provides the building blocks for the data structure of GEOS objects.
```

File: src/coreComponents/discretizationMethods/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/discretizationMethods/CMakeLists.txt]: What do you think of:
```suggestion
# Gives access to the discretization methods and their components for applications to the mesh.
```

File: src/coreComponents/events/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/events/CMakeLists.txt]: May need to be revised a bit

File: src/coreComponents/fieldSpecification/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/fieldSpecification/CMakeLists.txt]:
```suggestion
# Contains fields specification objects definition
```

File: src/coreComponents/fileIO/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/fileIO/CMakeLists.txt]:
```suggestion
# Contains I/O interfaces for the packing components and supported outputs and their python wrappers.
# Contains coupler for data exchange with CHOMBO.
```

File: src/coreComponents/finiteElement/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/finiteElement/CMakeLists.txt]: Would this suffice?
```suggestion
# Contains the interface & implementation of calculations for different types of finite element.
```

File: src/coreComponents/finiteVolume/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/finiteVolume/CMakeLists.txt]:
```suggestion
# Storing and access to the finite volume data. The main data structures of this package are the stencils: the connection values at the interface of the finite volumes.
```

File: src/coreComponents/functions/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/functions/CMakeLists.txt]:
```suggestion
# Contains classes for storing and computing arbitrary N-dimensional functions.
```

File: src/coreComponents/mainInterface/CMakeLists.txt
Comment (MelReyCG) [inline src/coreComponents/mainInterface/CMakeLists.txt]:
```suggestion
# Contains:
#  - GEOS `main` entry point.
#  - the class managing the operation flow of the problem being ran in GEOS.
#  - basic initialization and environment setup routines.
```

**Review by MelReyCG — COMMENTED** [review]
Some proposals

File: src/coreComponents/finiteElement/CMakeLists.txt
Comment (paloma-martinez) [reply src/coreComponents/finiteElement/CMakeLists.txt]: I'm not entirely okay with it, I suggest that

`# Contains the interface and implementation of finite element formulations.
`

File: src/coreComponents/constitutiveDrivers/CMakeLists.txt
Comment (paloma-martinez) [reply src/coreComponents/constitutiveDrivers/CMakeLists.txt]: I reworded it as:
 
 `# Contains fluid and rock constitutive model tests to calibrate constitutive model parameters to experimental data.`

File: src/coreComponents/finiteElement/CMakeLists.txt
Comment (paveltomin) [inline src/coreComponents/finiteElement/CMakeLists.txt]: is it called FEMKI anywhere else?

**Review by paveltomin — APPROVED** [review]

File: src/coreComponents/finiteElement/CMakeLists.txt
Comment (paloma-martinez) [reply src/coreComponents/finiteElement/CMakeLists.txt]: I found `FEMKI` term  in the sphinx documentation (developer guide), but I realize it is only used there, probably only for lighter writing. I'll remove it from the `CMakeList` doc.

File: src/coreComponents/constitutiveDrivers/CMakeLists.txt
Comment (MelReyCG) [reply src/coreComponents/constitutiveDrivers/CMakeLists.txt]: Very clear!

**Review by MelReyCG — APPROVED** [review]

**Comment by paveltomin** [comment]
@CusiniM, @andrea-borio, @castelletto1, @corbett5, @cssherman, @rrsettgast, @untereiner, @victorapm, and/or @wrtobin your review is required

**Review by victorapm — APPROVED** [review]

**Review by untereiner — COMMENTED** [review]
- For multi-line comments I would prefer using blocks as explained [here](https://cmake.org/cmake/help/v3.1/manual/cmake-language.7.html#id25). An example [here](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/AddFileDependencies.cmake?ref_type=heads)

- Also the specific block :
```
#[====================[.rst:

#]====================]
```
allows to generate sphinx documentation. Look [here](https://discourse.cmake.org/t/how-does-cmake-docs-use-sphinx-to-parse-comments-in-cmake-files/8593) for more details.

- Also I suggest to add [spdx ](https://spdx.dev/) info to the cmake files as well. Something like
```
SPDX-License-Identifier: LGPL-2.1-only
SPDX-FileCopyrightText: © 2016-2024 Lawrence Livermore National Security LLC
SPDX-FileCopyrightText: © 2018-2024 Totalenergies, S.A
SPDX-FileCopyrightText: © 2018-2024 The Board of Trustees of the Leland Stanford Junior University
SPDX-FileCopyrightText: © 2023-2024 Chevron
SPDX-FileCopyrightText: © 2019-2024 GEOS/GEOSX Contributors
```

**Comment by paloma-martinez** [comment]
@untereiner 
The generation of Sphinx doc from these files is planned for another PR (in progress), along with the formatting aspects that come with it.

For now, would you be okay with this kind of format for the files headers ?

 ```
 #[[ SPDX-License-Identifier: LGPL-2.1-only
 * Copyright (c) 2016-2024 Lawrence Livermore National Security LLC
 * Copyright (c) 2018-2024 Total, S.A
 * Copyright (c) 2018-2024 The Board of Trustees of the Leland Stanford Junior University
 * Copyright (c) 2023-2024 Chevron
 * Copyright (c) 2019-     GEOS/GEOSX Contributors
 * All rights reserved

See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.]]
#--------------------------------------------------------------------------------------------------

#[[
Package: codingUtilities

Contains definition of runtime types. Also contains common traits.]]
```

**Comment by untereiner** [comment]
Why modifying twice the same files ?
What do you think of :
```
# SPDX-License-Identifier: LGPL-2.1-only
# SPDX-FileCopyrightText: © 2016-2024 Lawrence Livermore National Security LLC
# SPDX-FileCopyrightText: © 2018-2024 Totalenergies, S.E
# SPDX-FileCopyrightText: © 2018-2024 The Board of Trustees of the Leland Stanford Junior University
# SPDX-FileCopyrightText: © 2023-2024 Chevron
# SPDX-FileCopyrightText: © 2019-2024 GEOS/GEOSX Contributors

#[====================[
Package: codingUtilities

Contains definition of runtime types. Also contains common traits.
#]====================]

```

**Comment by paloma-martinez** [comment]
> Why modifying twice the same files ? What do you think of :
> 
> ```
> # SPDX-License-Identifier: LGPL-2.1-only
> # SPDX-FileCopyrightText: © 2016-2024 Lawrence Livermore National Security LLC
> # SPDX-FileCopyrightText: © 2018-2024 Totalenergies, S.E
> # SPDX-FileCopyrightText: © 2018-2024 The Board of Trustees of the Leland Stanford Junior University
> # SPDX-FileCopyrightText: © 2023-2024 Chevron
> # SPDX-FileCopyrightText: © 2019-2024 GEOS/GEOSX Contributors
> 
> #[====================[
> Package: codingUtilities
> 
> Contains definition of runtime types. Also contains common traits.
> #]====================]
> ```

I'm OK with the SPDX part.
At the moment, the sphinx package moderncmakedomain does not give the result we are expecting, so I would prefer delaying the modifications linked to it to the future PR, and look for the best solution for documentation.

**Review by untereiner — APPROVED** [review]

**Comment by paveltomin** [comment]
needs review/reviews from @CusiniM, @andrea-borio, @castelletto1, @corbett5, @cssherman, @rrsettgast, and/or @wrtobin

File: src/coreComponents/denseLinearAlgebra/CMakeLists.txt
Comment (CusiniM) [inline src/coreComponents/denseLinearAlgebra/CMakeLists.txt]:
```suggestion
Contains dense linear algebra functions and interfaces to BLAS and LAPACK packages.
```

**Review by CusiniM — APPROVED** [review]

**Comment by CusiniM** [comment]
codeowners approvals are missing. Please move back to the merge queue once the PR is fully approved.

**Comment by paloma-martinez** [comment]
Hi @corbett5 @cssherman @rrsettgast and/or @wrtobin, could you please review this PR as code owners

**Review by corbett5 — APPROVED** [review]
LGTM

**Comment by MelReyCG** [comment]
I removed `ci: run code coverage` and `ci: run CUDA builds`, they encountered no issue.

**Comment by MelReyCG** [comment]
Just to notify people here, @paloma-martinez will be off for 4 months.
- I'm adding a brief for `src/coreComponents/unitTests`
- We have only 2 briefs left to review here:
  * `src/coreComponents/functions`
  * `src/coreComponents/unitTests`

**Review by wrtobin — APPROVED** [review]

**Pull request merged**
